How to build and find stable branch

gautam1168 opened the pull request.

gautam1168 committed:
How to build and find stable branch

gautam1168 commented:
@mattgodbolt I am opening this pull request primarily to answer the question you asked about which branch to build. I have answered that in the PR in the last section. But I will repeat again here. v8 releases are tied to chrome releases. And every time a chrome release happens a new branch is created in v8 for that release. The name of this branch can be found for  linux by looking up the `true branch` column in this page: https://omahaproxy.appspot.com/ and prefixing it with `chromium`. So for example the branch right now that is stable on linux is `chromium/5563`

I built this docker image today on my local machine but I didn't see the build.sh file inside the /root directory. I am not sure why that is the case. But Ill figure it out and try running the build.sh script. However one line would require a change in the build.sh file. Namely
```
if echo ${VERSION} | grep 'trunk'; then
	VERSION="trunk-$(date +%Y%m%d)"
	BRANCH="main"
else
	BRANCH="branch-heads/${VERSION}"
fi
```

will have to be changed to 
```
if echo ${VERSION} | grep 'trunk'; then
	VERSION="trunk-$(date +%Y%m%d)"
	BRANCH="main"
else
	BRANCH="chromium/${VERSION}"
fi
```

As this blog post is saying https://v8.dev/docs/release-process. the `branch-heads/x.y` versions are just beta branches that get promoted to stable in v8. While the `chromium/number` branches are the v8 version existing in the latest versions of chrome. I am not sure which one of these you would like to go with. 

I will try to figure out why the build.sh is not showing up in my container, as well as write a small script to get the latest of these numbers out of omahaproxy automatically. Meanwhile please tell me if you know which one of these stable versions should be put up on ce.

jakobkummerow commented:
When our documentation says "use the tip of the branch" (in this case `branch-heads/11.1`), we mean it.

> for example the branch right now that is stable on linux is `chromium/5563`

That is incorrect. The V8 `chromium/5563` branch was created for the Chrome 111.0.5563.0 Canary. It served as the base point for the 11.1 branch, which was stabilized (by merging fixes to it) along with Chrome 111. The Chrome 111.0.5563.64 stable build contains the V8 version that was the tip of `branch-heads/11.1` when that Chrome build was built (a few days before its release).

> I will [...] write a small script to get the latest of these numbers out of omahaproxy automatically

Be advised that omahaproxy is (unfortunately) [going away in a few weeks](https://groups.google.com/a/chromium.org/g/chromium-dev/c/uH-nFrOLWtE).

gautam1168 commented:
@jakobkummerow thankyou for helping 😊 🙇.

@mattgodbolt so to get the latest version we can just list the branches that start with `branch-heads` and take the latest one. This simplifies things.

gautam1168 committed:
Update README.md

gautam1168 commented:
@mattgodbolt @jakobkummerow I have updated the readme with a command at the end to find the latest version. I used chat-gpt to generate that command. @mattgodbolt note that the change I mentioned in my first comment is not needed thanks to the information that @jakobkummerow provided. I am not sure how ce figures out which version it should give the build command, but you can get that version using the command in the last section of the readme in this PR.

mattgodbolt commented on README.md:
````diff
@@ -1,4 +1,69 @@
+To get the latest version of v8 you can list the branches that start with `branch-heads`. You can use this command to get the latest version:
+```
+git branch -a | grep remotes/branch-heads | awk -F/ '{print $NF}' | sort -t. -k1,1n
+r -k2,2nr -k3,3nr | head -n1
````
What is the `r` command? I've never heard of it before? Should this be part of the previous line?

mattgodbolt replied:
If it's a long line (which it would be if this is supposed to be a single line) I'd suggest breaking the line and using a `\`

```
git branch -a | grep remotes/branch-heads | awk -F/ '{print $NF}' \
    | sort -t. -k1,1nr -k2,2nr -k3,3nr | head -n1
```

gautam1168 replied:
Yes this is a paste problem, because I copied from my windows terminal and that thing is weird about copy pasting stuff. The command is is actually a single line like this:
```
 git branch -a | grep remotes/branch-heads | awk -F/ '{print $NF}' | sort -t. -k1,1nr -k2,2nr -k3,3nr | head -n1
```

mattgodbolt commented on README.md:
````diff
@@ -1,4 +1,69 @@
+
+You can then open a shell inside this container using:
+```
+docker exec -it cfdab51b3818 fish
````
There is no `fish` inside this container. (It's my shell of choice too, but I don't put it into general purpose things; it's enough having to expalin to folks how all this stuff works without giving them a new shell to contend with too!)

gautam1168 replied:
Yes that makes sense. I see what happened here. I used my own image instead of the this container. I'll change this to sh or bash.

mattgodbolt commented on README.md:
````diff
@@ -1,4 +1,69 @@
+
+Once you have this, you can run a container with this image using the following command:
+```
+docker run -d -t d8:trunk
````
I really wouldn't run it this way myself. There's no need to have a detached session running - this is like running a UNIX shell as a `daemon` and then connectint to it, with all the attendant problems of remembering to stop and remove it after.

Instead if you run:

```sh
docker run --rm -it d8:trunk bash
```

You wlil get dropped into a fresh, temporary container, running `bash` interactively. The container is removed as soon as that bash shell exits. Then there's no need to `docker exec` or `docker ps`

gautam1168 replied:
This makes sense. I use the docker desktop app and this lets me use the UI to see the running container and open a terminal from there if I choose. However, I know that my choice of windows as my working machine is a weird one and I will update it to this version.

mattgodbolt reviewed (changes requested):
Thanks! i appreciate the documentation; though I have a number of comments about how to improve it, and a query about the part at the end.

gautam1168 committed:
Fix command to list version and modify shell command

gautam1168 committed:
Update README.md

gautam1168 commented:
I have updated the PR to add the changes as you suggested. Also I have fixed the error in the last command. That was a paste error because of my windows terminal.

Although I have updated the shell command as you suggested, I use a permanent container instead of a temporary one. Because I just restart that container and do stuff with the d8 executable and then stop it again. But I guess anyone who is going to set this much up can figure out how to make a persistent container as well.

gautam1168 committed:
Add linebreak in command

mattgodbolt reviewed (approved):
Thanks!

gautam1168 commented:
I will update my PR https://github.com/compiler-explorer/compiler-explorer/pull/4721 today to use the  build made by this container.

mattgodbolt commented:
Thanks @gautam1168 for your patience in this.

gautam1168 commented:
I am over here thinking you must be getting annoyed by the slow pace at which I'm doing things. I assure you I myself am thoroughly enjoying this experience. And I love it every time @jakobkummerow responds to correct something I'm doing as well. I hope I don't end up annoying him and driving him away either 😅. I am fascinated by v8 and I welcome any opportunity to learn more about it.
